features/messages/impl/src/main/kotlin/io/element/android/features/messages/impl/report/ReportMessagePresenter.kt

```diff
@@ -0,0 +1,99 @@
+
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.MutableState
+import androidx.compose.runtime.getValue
+import androidx.compose.runtime.mutableStateOf
+import androidx.compose.runtime.remember
+import androidx.compose.runtime.rememberCoroutineScope
+import androidx.compose.runtime.saveable.rememberSaveable
+import androidx.compose.runtime.setValue
+import dagger.assisted.Assisted
+import dagger.assisted.AssistedFactory
+import dagger.assisted.AssistedInject
+import io.element.android.libraries.architecture.Async
+import io.element.android.libraries.architecture.Presenter
+import io.element.android.libraries.architecture.executeResult
+import io.element.android.libraries.core.coroutine.CoroutineDispatchers
+import io.element.android.libraries.designsystem.utils.SnackbarDispatcher
+import io.element.android.libraries.designsystem.utils.SnackbarMessage
+import io.element.android.libraries.matrix.api.core.EventId
+import io.element.android.libraries.matrix.api.core.UserId
+import io.element.android.libraries.matrix.api.room.MatrixRoom
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.launch
+import io.element.android.libraries.ui.strings.R as StringR
+
+class ReportMessagePresenter @AssistedInject constructor(
+    private val room: MatrixRoom,
+    @Assisted private val inputs: Inputs,
+    private val snackbarDispatcher: SnackbarDispatcher,
+    private val dispatchers: CoroutineDispatchers,
+) : Presenter<ReportMessageState> {
+
+    data class Inputs(
+        val eventId: EventId,
+        val senderId: UserId,
+    )
+
+    @AssistedFactory
+    interface Factory {
+        fun create(inputs: Inputs): ReportMessagePresenter
+    }
+
+    @Composable
+    override fun present(): ReportMessageState {
+        val coroutineScope = rememberCoroutineScope()
+        var reason by rememberSaveable { mutableStateOf("") }
+        var blockUser by rememberSaveable { mutableStateOf(false) }
+        var result: MutableState<Async<Unit>> = remember { mutableStateOf(Async.Uninitialized) }
+
+        fun handleEvents(event: ReportMessageEvents) {
+            when (event) {
+                is ReportMessageEvents.UpdateReason -> reason = event.reason
+                ReportMessageEvents.ToggleBlockUser -> blockUser = !blockUser
+                ReportMessageEvents.Report -> coroutineScope.report(inputs.eventId, inputs.senderId, reason, blockUser, result)
+                ReportMessageEvents.ClearError -> result.value = Async.Uninitialized
+            }
+        }
+
+        return ReportMessageState(
+            reason = reason,
+            blockUser = blockUser,
+            result = result.value,
+            eventSink = ::handleEvents
+        )
+    }
+
+    private fun CoroutineScope.report(
+        eventId: EventId,
+        userId: UserId,
+        reason: String,
+        blockUser: Boolean,
+        result: MutableState<Async<Unit>>,
+    ) = launch(dispatchers.main) {
```

Comment:
Is this the best dispatcher for this use case?

Reply:
`room.reportContent` later changes the dispatcher to `IO` so it should be fine. 

In any case, I made this change because the tests were acting weird with loading states, but I just tried removing the custom dispatchers altogether and apparently it works fine without them, so I think I'll just remove them and simplify the code.